exporter: promethize counter names

pereman2 opened the pull request:
This fix assures that metrics coming from daemons perf counters have the same name if they are exported by the Prometheus manager module or by the Ceph exporter. 

Fixes: https://bugzilla.redhat.com/show_bug.cgi?id=2186557

jmolmo commented:
Output with the new code. the "_", "+" has been properly replaced.

```
(standard input)752:ceph_mds_mem_cap_minus{ceph_daemon="mds.a"} 0
(standard input)753:# HELP ceph_mds_mem_cap_plus Capabilities added
(standard input)754:# TYPE ceph_mds_mem_cap_plus counter
(standard input)755:ceph_mds_mem_cap_plus{ceph_daemon="mds.a"} 0
(standard input)756:# HELP ceph_mds_mem_dir Directories
(standard input)757:# TYPE ceph_mds_mem_dir gauge
(standard input)758:ceph_mds_mem_dir{ceph_daemon="mds.a"} 12
(standard input)759:# HELP ceph_mds_mem_dir_minus Directories closed
```

idryomov commented:
> Output with the new code. the "_", "+" has been properly replaced.

Only because `+` happens to be last character in this example.

pereman2 commented on src/exporter/DaemonMetricCollector.cc:
```diff
@@ -80,7 +81,34 @@ std::string boost_string_to_std(boost::json::string js) {
+
+  size_t double_colon_pos = name.find("::");
+  if (double_colon_pos != std::string::npos) {
+    name.replace(double_colon_pos, 2, "_");
```
true, I'll change that.

pereman2 commented on src/exporter/DaemonMetricCollector.cc:
```diff
@@ -80,7 +81,34 @@ std::string boost_string_to_std(boost::json::string js) {
+
+  // just in case because there are some bluestore metrics which include a single ":"
+  if (name.find(":") != std::string::npos) {
+    boost::replace_all(name, ":", "_");
```
as the comment says, I saw a single colon around some bluestore metrics so I added that in just in case.

idryomov replied:
Then `promethize` in `src/pybind/mgr/prometheus/module.py` needs to updated to the same.  There can't be "just in case" here, both methods should do exactly the same thing.

epuertat replied:
Yup, it's important to keep consistency with the existing metric names (even if the current approach was not the best possible). Otherwise, this will generate new time-series, and the existing Grafana Dashboards might have missing data.

Maybe those bluestore metrics are not exported due to their priority level?

pereman2 commented:
I made the code replace all + with `_plus` just in case. Expecting a + as an infix will be a weird case so I don't think is something we should care about for now.

Another thing we could add is replace all invalid characters from the prometheus name regex they provide officially to underscores or something simliar but I don't feel like we need to reach that far for the current quick fix.

epuertat commented on src/exporter/DaemonMetricCollector.cc:
```diff
@@ -80,7 +81,23 @@ std::string boost_string_to_std(boost::json::string js) {
 std::string quote(std::string value) { return "\"" + value + "\""; }
 
-bool is_hyphen(char ch) { return ch == '-'; }
+void promethize(std::string &name) {
```
I'll approve whatever this function does if you unit-test it by taking a dump of all (current) perfcounters (or a neatly chosen subset of those) and this outputs the exact same metric names as the [`mgr/prometheus` exporter](https://prometheus.io/docs/concepts/data_model/#metric-names-and-labels).

It should be mandatory to write unit tests for pure functions.

pereman2 replied:
I've decided to copy all the metrics I could gather from a `vstart.sh` cluster `MON=1 OSD=1 MDS=1 FS=1 NFS=1 RGW=1` as parameters. With  ~600 metrics I tested the function based on how the `prometheus/module.py` performs it. I hope this helps.

epuertat replied:
👍🏼

epuertat commented on src/exporter/DaemonMetricCollector.cc:
```diff
@@ -138,8 +155,8 @@ void DaemonMetricCollector::dump_asok_metrics() {
         std::string counter_name_init =  {counter.key().begin(), counter.key().end()};
-        std::string counter_name = "ceph_" + labeled_perf_group + "_" + counter_name_init;
-        std::replace_if(counter_name.begin(), counter_name.end(), is_hyphen, '_');
+        std::string counter_name = perf_group + "_" + counter_name_init;
+        promethize(counter_name);
```
Silly question: what about just `promethizing` the changing suffix?

 ```suggestion
        std::string counter_name = std::format("{}_{}_{}", CEPH_PREFIX, perf_group, promethize(counter_name_init));
```
If `perf_group` needed sanitization too, it could be done in the outer for-loop.

pereman2 replied:
I didn't add the `ceph_` prefix before calling `promethize` because I wanted the function to be as close to the python implementation to add unit tests faster.

epuertat replied:
Ok, makes sense!

epuertat reviewed (approved):
LGTM! Thank you, @pereman2 !

BTW, I see no "Fixes:...". Shouldn't this be linked to a bug report?

epuertat commented on src/test/exporter/promethize_data.h:
```diff
@@ -0,0 +1,652 @@
+  {"mds_log.segexg", "ceph_mds_log_segexg"},
+  {"mds_log.segtrm", "ceph_mds_log_segtrm"},
+  {"mds_mem.cap", "ceph_mds_mem_cap"},
+  {"mds_mem.cap+", "ceph_mds_mem_cap_plus"},
+  {"mds_mem.cap-", "ceph_mds_mem_cap_minus"},
+  {"mds_mem.dir", "ceph_mds_mem_dir"},
+  {"mds_mem.dir+", "ceph_mds_mem_dir_plus"},
+  {"mds_mem.dir-", "ceph_mds_mem_dir_minus"},
+  {"mds_mem.dn", "ceph_mds_mem_dn"},
+  {"mds_mem.dn+", "ceph_mds_mem_dn_plus"},
+  {"mds_mem.dn-", "ceph_mds_mem_dn_minus"},
+  {"mds_mem.heap", "ceph_mds_mem_heap"},
+  {"mds_mem.ino", "ceph_mds_mem_ino"},
+  {"mds_mem.ino+", "ceph_mds_mem_ino_plus"},
+  {"mds_mem.ino-", "ceph_mds_mem_ino_minus"},
```
BTW, [love it](https://jenkins.ceph.com/job/ceph-pull-requests/113822/testReport/projectroot.src.test/exporter/unittest_exporter/)! 6 ms and ~640 different metric names checked. That's _the_ unit test:
```
Running main() from gmock_main.cc
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from Exporter
[ RUN      ] Exporter.promethize
[       OK ] Exporter.promethize (6 ms)
[----------] 1 test from Exporter (6 ms total)
```

My only concern is... is this the only unit test for this daemon? Aren't there other (pure) functions that would benefit from this hardening for regression/feature parity testing? Beyond this, what other kind of component or integration testing exists for ceph-exporter?

jmolmo replied:
For sure a more complete test coverage will be useful, and probably we need to set as priority.  However, I suggest to be focused in this pr just in the problem to resolve.

pereman2 replied:
I'll add the memento mori as I have to update the commit message.

As @jmolmo said, we should stick with the current purpose of this PR and @avanthakkar can work on on extending the testing suite.

I think there is no testing at all related to ceph-exporter, maybe avan can correct me here.

epuertat replied:
Sure, I was just to raise awareness on the fact the we backported a new feature to an stable release with barely any automated testing (context: part of the discussion during the CLT meeting was about possible lack of teuthology testing on the monitoring side being key to this oversight).

jmolmo commented:
The windows tests fail with an error in one of the exporter files, the weird thing is that the offending line has not been modified.... and the line was introduced 9 monthsd ago ...
`/home/ubuntu/ceph/src/exporter/DaemonMetricCollector.cc:245:27: error: ‘_SC_CLK_TCK’ was not declared in this scope; did you mean ‘CLK_TCK’?`

@pereman2 what do you think?

idryomov commented:
> The windows tests fail with an error in one of the exporter files, the weird thing is that the offending line has not been modified.... and the line was introduced 9 monthsd ago ...
> `/home/ubuntu/ceph/src/exporter/DaemonMetricCollector.cc:245:27: error: ‘_SC_CLK_TCK’ was not declared in this scope; did you mean ‘CLK_TCK’?`

It's failing to build the test, not the daemon.  I think the failure appeared because ceph-exporter isn't built for Windows but with this PR you are trying to build its test binary.  You would need to disable that in `src/test/CMakeLists.txt` by moving `add_subdirectory(exporter)` line into the "Not available on Windows for the time being" section.

Cc @petrutlucian94 to keep me honest.

idryomov commented on src/exporter/DaemonMetricCollector.cc:
```diff
@@ -80,7 +80,22 @@ std::string boost_string_to_std(boost::json::string js) {
+    name[name.size() - 1] = '_';
+    name += "minus";
+  } 
+  boost::replace_all(name, "-", "_");
```
Can `minus` handling be moved to the top of the method and this (fourth!) iteration be folded into `should_be_underscore` one to reduce the number of iterations and so that `boost::replace_all` is used only for "complex" substitutions?

idryomov commented on src/test/exporter/promethize_data.h:
```diff
@@ -0,0 +1,655 @@
+#include <string>
+#include <vector>
+
+using namespace std;
```
`using namespace` directive in header files is a bad idea and should be avoided, especially if it's for `std`.

Taking a step back, why is this in a header file to begin with? Do you expect promethize_data vector to be used by something other than Exporter.promethize test case?

pereman2 replied:
Removed the using std. 

I moved this to a header file to remove potential bloat for the unittest file of the exporter.

idryomov replied:
> I moved this to a header file to remove potential bloat for the unittest file of the exporter.

Since it's getting compiled for the unittest (in the same translation unit, and _only_ in that translation unit), I don't think that is a valid reason for creating a header file with `#pragma once`, etc.  Moving over a definition like in the source file takes one or two keystrokes in any decent editor.

idryomov commented on src/exporter/DaemonMetricCollector.cc:
```diff
@@ -80,7 +80,22 @@ std::string boost_string_to_std(boost::json::string js) {
 std::string quote(std::string value) { return "\"" + value + "\""; }
 
-bool is_hyphen(char ch) { return ch == '-'; }
+void promethize(std::string &name) {
```
Let's add "Must be kept in sync with promethize() in src/pybind/mgr/prometheus/module.py" comment here and a corresponding comment in `src/pybind/mgr/prometheus/module.py`.

petrutlucian94 commented:
> > The windows tests fail with an error in one of the exporter files, the weird thing is that the offending line has not been modified.... and the line was introduced 9 monthsd ago ...
> > `/home/ubuntu/ceph/src/exporter/DaemonMetricCollector.cc:245:27: error: ‘_SC_CLK_TCK’ was not declared in this scope; did you mean ‘CLK_TCK’?`
> 
> It's failing to build the test, not the daemon. I think the failure appeared because ceph-exporter isn't built for Windows but with this PR you are trying to build its test binary. You would need to disable that in `src/test/CMakeLists.txt` by moving `add_subdirectory(exporter)` line into the "Not available on Windows for the time being" section.
> 
> Cc @petrutlucian94 to keep me honest.

That's correct, please skip this test on Windows. Thanks!

idryomov commented on src/exporter/DaemonMetricCollector.cc:
```diff
@@ -1,5 +1,6 @@
 #include "DaemonMetricCollector.h"
 
+#include <boost/algorithm/string/replace.hpp>
```
This include is no longer needed.

idryomov commented on src/exporter/util.h:
```diff
@@ -20,3 +20,6 @@ class BlockTimer {
 std::string read_file_to_string(std::string path);
 std::string get_hostname(std::string path);
+
+// Must be kept in sync with promethize() in src/pybind/mgr/prometheus/module.py
```
Move this comment to the definition -- it's much more likely to be noticed there.  If someone tweaks the body without needing to tweak the signature chances are they won't even open the header file.

idryomov replied:
And please add a corresponding comment to `src/pybind/mgr/prometheus/module.py` here, in the same commit.

idryomov commented on src/test/exporter/CMakeLists.txt:
```diff
@@ -0,0 +1,12 @@
+add_executable(unittest_exporter
+  test_exporter.cc
+  "${CMAKE_SOURCE_DIR}/src/exporter/DaemonMetricCollector.cc"
```
Does `DaemonMetricCollector.cc` still need to be built here?

jmolmo commented:
jenkins test make check

jmolmo commented:
jenkins test make check arm64

jmolmo commented:
jenkins test make check arm64

idryomov commented on src/test/exporter/test_exporter.cc:
```diff
@@ -0,0 +1,662 @@
+#include "gtest/gtest.h"
+#include "exporter/util.h"
+
+#include <string>
```
`<vector>` include appears to have been lost in the move ;)  Strictly speaking, also need to include `<utility>` here for `std::pair`.

idryomov commented on src/exporter/DaemonMetricCollector.cc:
```diff
@@ -120,12 +118,14 @@ void DaemonMetricCollector::dump_asok_metrics() {
+      json_object perf_group_object = perf_group_item.value().as_object();
+      auto counters = perf_group_object["counters"].as_object();
+      auto counters_labels = perf_group_object["labels"].as_object();
+      auto labeled_perf_group_counters =
```
Kudos for getting rid of "labeled" in most of the names here!  But this variable still has it for some reason, despite the counters in question not necessarily being labeled.  I'd suggest "counters_values" for symmetry with "counters_labels".

avanthakkar replied:
Actually, this variable specifies the labels field we need to fetch from the dump command.

idryomov replied:
No, it doesn't.  I think you are confusing it with `counters_labels`.

avanthakkar commented:
jenkins test dashboard cephadm

avanthakkar commented:
jenkins test dashboard cephadm

jmolmo commented:
jenkins test dashboard cephadm

avanthakkar commented:
jenkins test make check

pereman2 commented:
@idryomov I've fixed the last comments from you. 

@avanthakkar can you give a brief on why this commit was added ? https://github.com/ceph/ceph/pull/51069/commits/4998f85f3b98fedd9632debaa033b7e5d7fd6963 I've removed it completely as you didn't give any explanations on the commit plus it didn't follow naming conventions of commit messages.

idryomov reviewed (approved):
Approving based on

> For sure a more complete test coverage will be useful, and probably we need to set as priority. However, I suggest to be focused in this pr just in the problem to resolve.

> As @jmolmo said, we should stick with the current purpose of this PR and @avanthakkar can work on on extending the testing suite.

@avanthakkar Can you link the respective tracker ticket here?

avanthakkar commented:
jenkins test dashboard

jmolmo commented:
Ceph dashboard test failures not related with this change
